### PR TITLE
[MXNet] Add lab-04-3 and lab-12-4

### PR DESCRIPTION
**Pull request by sxjscience**
Add LinearRegression + Basic RNN

### CODE REVIEW

**Comment by sxjscience** [comment]
I'm referring to the Keras lab files for the implementation.

File: mxnet/mxlab-04-3-file_input_linear_regression.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+# 2. Build Linear Regression Model
+data = mx.sym.Variable("data")
+target = mx.sym.Variable("target")
+pred = mx.sym.FullyConnected(data=data, num_hidden=1, name='pred')
+loss = mx.sym.mean(mx.sym.square(target - pred))
+loss = mx.sym.make_loss(loss)
+
+
+# 3. Build the network
```
Comment (hunkim) [inline mxnet/mxlab-04-3-file_input_linear_regression.py]: What is this part? Do we need to bind all the time? It looks a bit complicated for a simple regression model.

File: mxnet/mxlab-04-3-file_input_linear_regression.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+# 4. Train the model
+y_data_nd = nd.array(y_data, ctx=mx.gpu())
+for idx in range(2000):
+    total_batch_num = int(sample_num / batch_size)
+    for batch_id in range(total_batch_num):
+        x_batch = x_data[batch_id*batch_size:(batch_id+1)*batch_size , :]
+        y_batch = y_data[batch_id*batch_size:(batch_id+1)*batch_size , :]
+        data_batch = mx.io.DataBatch(data=[mx.nd.array(x_batch)],
+                                     label=[mx.nd.array(y_batch)])
+        net.forward(data_batch, is_train=True)
```
Comment (hunkim) [inline mxnet/mxlab-04-3-file_input_linear_regression.py]: Always we need to do 1) forward, backward, and update? These three are very common, so do you have something like `net.train(data_batch)`?

File: mxnet/mxlab-04-3-file_input_linear_regression.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        net.backward()
+        net.update()
+    test_net.forward(mx.io.DataBatch(data=[mx.nd.array(x_data)], label=None), is_train=False)
+    pred_nd = test_net.get_outputs()[0]
```
Comment (hunkim) [inline mxnet/mxlab-04-3-file_input_linear_regression.py]: This is also interesting. After forward, we need to get_outputs, right?

File: mxnet/mxlab-11-5-mnist_cnn_ensemble_layers.py [hunk 4]
```diff
@@ -1,5 +1,4 @@
 # Lab 11 MNIST and Deep learning CNN
-# https://www.tensorflow.org/tutorials/layers
```
Comment (hunkim) [inline mxnet/mxlab-11-5-mnist_cnn_ensemble_layers.py]: Good idea!

File: mxnet/mxlab-12-4-rnn_deep_prediction.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        lstm1 = mx.rnn.LSTMCell(num_hidden=5, prefix="lstm1_")
+        lstm2 = mx.rnn.LSTMCell(num_hidden=10, prefix="lstm2_")
+    L1, _ = lstm1.unroll(length=seq_len, inputs=data, merge_outputs=True,
+                         layout="TNC")  # Shape: (T, N, 5)
+    L1 = mx.sym.Dropout(L1, p=0.2)  # Shape: (T, N, 5)
+    _, L2_states = lstm2.unroll(length=seq_len, inputs=L1, merge_outputs=True,
+                                layout="TNC")  # Shape: (T, N, 10)
+    L2 = mx.sym.reshape(L2_states[0], shape=(-1, 0), reverse=True)  # Shape: (T * N, 10)
+    pred = mx.sym.FullyConnected(L2, num_hidden=1, name="pred")
+    pred = mx.sym.LinearRegressionOutput(data=pred, label=target)
```
Comment (hunkim) [inline mxnet/mxlab-12-4-rnn_deep_prediction.py]: Can we just use this for lab-04, linear regression?

File: mxnet/mxlab-04-3-file_input_linear_regression.py (hunk 3 above)
Comment (sxjscience) [inline mxnet/mxlab-04-3-file_input_linear_regression.py]: Yes, the current API requires using get_outputs() to gather the outputs. This is mainly designed for multi-GPU training.

For example, we can set the context of `mx.mod.Module` to `context=[mx.gpu(0), mx.gpu(1)]`. Inside the module object, two executors will be created separately on two GPUs. When the user calls `net.forward()`, the two executors will be forwarded separately. And the `net.get_outputs()` needs to be called to synchronize the result to the first GPU.

File: mxnet/mxlab-12-4-rnn_deep_prediction.py (hunk 5 above)
Comment (sxjscience) [inline mxnet/mxlab-12-4-rnn_deep_prediction.py]: Yes, we can use this for LinearRegression. Do I need to change it? The LinearRegressionOutput symbol is something special. In the forward pass, it behaves like an identity function while in the backward pass, it passes back the gradient calculated using the squared loss.

File: mxnet/mxlab-04-3-file_input_linear_regression.py (hunk 2 above)
Comment (sxjscience) [inline mxnet/mxlab-04-3-file_input_linear_regression.py]: @hunkim I've changed to `net.forward_backward(data_batch)`. Currently there is no short_cut for `net.train(data_batch)`.

File: mxnet/mxlab-12-4-rnn_deep_prediction.py (hunk 5 above)
Comment (hunkim) [inline mxnet/mxlab-12-4-rnn_deep_prediction.py]: It's up to you.

File: mxnet/mxlab-04-3-file_input_linear_regression.py [hunk 6]
```diff
@@ -52,8 +52,7 @@
         y_batch = y_data[batch_id*batch_size:(batch_id+1)*batch_size , :]
         data_batch = mx.io.DataBatch(data=[mx.nd.array(x_batch)],
                                      label=[mx.nd.array(y_batch)])
-        net.forward(data_batch, is_train=True)
-        net.backward()
+        net.forward_backward(data_batch)
```
Comment (hunkim) [inline mxnet/mxlab-04-3-file_input_linear_regression.py]: Looks better. Still need to update. :-)

File: mxnet/mxlab-04-3-file_input_linear_regression.py (hunk 1 above)
Comment (sxjscience) [inline mxnet/mxlab-04-3-file_input_linear_regression.py]: Yes, the lines above only construct the symbol. And we will need to bind the symbol to an executor in order to run the code.

File: mxnet/mxlab-04-3-file_input_linear_regression.py (hunk 6 above)
Comment (sxjscience) [inline mxnet/mxlab-04-3-file_input_linear_regression.py]: I plan to PR a shortcut function in MXNet master.

**Comment by hunkim** [comment]
OK. I'll merge!

File: mxnet/mxlab-04-3-file_input_linear_regression.py (hunk 6 above)
Comment (hunkim) [inline mxnet/mxlab-04-3-file_input_linear_regression.py]: How about `mod.fit`?
